File: app/src/main/java/cn/alvkeke/dropto/storage/DataLoader.java
package cn.alvkeke.dropto.storage;

import android.content.Context;
import android.util.Log;

import java.util.ArrayList;

import cn.alvkeke.dropto.data.Category;

public class DataLoader {

    private static DataLoader instance = null;

    private DataLoader() { }

    public static DataLoader getInstance() {
        if (instance != null)
            return instance;

        instance = new DataLoader();
        return instance;
    }

    private ArrayList<Category> categories = null;

    private void loadCategories(Context context) {
        if (categories == null)
            categories = new ArrayList<>();
        try (DataBaseHelper helper = new DataBaseHelper(context)) {
            helper.start();
            helper.queryCategory(-1, categories);
            helper.finish();
        } catch (Exception ex) {
            Log.e(this.toString(), "Failed to retrieve category data from database");
        }
    }

    public ArrayList<Category> getCategories(Context context) {
        loadCategories(context);
        return categories;
    }

    public Category findCategory(Context context, long id) {
        loadCategories(context);

        for (Category c : categories) {
            if (c.getId() == id)
                return c;
        }
        return null;
    }

    public boolean loadCategoryNotes(Context context, Category category) {
        if (category.isInitialized())
            return true;
        try (DataBaseHelper helper = new DataBaseHelper(context)) {
            helper.start();
            helper.queryNote(-1, category.getId(), category.getNoteItems());
            category.setInitialized(true);
            helper.finish();
        } catch (Exception ex) {
            return false;
        }
        return true;
    }

}

File: app/src/main/java/cn/alvkeke/dropto/storage/DataLoader.kt
package cn.alvkeke.dropto.storage

import android.content.Context
import android.util.Log
import cn.alvkeke.dropto.data.Category

object DataLoader {

    @JvmStatic
    var categories: ArrayList<Category> = ArrayList()
        private set

    @JvmStatic
    fun loadCategories(context: Context): ArrayList<Category> {

        try {
            DataBaseHelper(context).use { helper ->
                helper.start()
                helper.queryCategory(-1, categories)
                helper.finish()
            }
        } catch (_: Exception) {
            Log.e(this.toString(), "Failed to retrieve category data from database")
        }

        return categories
    }

    @JvmStatic
    fun findCategory(id: Long): Category? {
        for (c in categories) {
            if (c.id == id) return c
        }
        return null
    }

    @JvmStatic
    fun loadCategoryNotes(context: Context, category: Category): Boolean {
        if (category.isInitialized)
            return true

        try {
            DataBaseHelper(context).use { helper ->
                helper.start()
                helper.queryNote(-1, category.id, category.noteItems)
                category.isInitialized = true
                helper.finish()
            }
        } catch (e: Exception) {
            Log.e(TAG, "${e.printStackTrace()}")
            return false
        }
        return true
    }

}
